gh-141004: Document remaining `pyport.h` utility macros (#144279)
python/cpython: Doc/c-api/intro.rst

   .. versionadded:: 3.3

.. c:macro:: Py_ALIGNED(num)

   Specify alignment to *num* bytes on compilers that support it.

   Consider using the C11 standard ``_Alignas`` specifier over this macro.

.. c:macro:: Py_ARITHMETIC_RIGHT_SHIFT(type, integer, positions)

   Similar to ``integer >> positions``, but forces sign extension, as the C
   standard does not define whether a right-shift of a signed integer will
   perform sign extension or a zero-fill.

   *integer* should be any signed integer type.
   *positions* is the number of positions to shift to the right.

   Both *integer* and *positions* can be evaluated more than once;

>    consequently, avoid directly passing a function call or some other
>    expensive operation to this macro. Instead, store the result as a
>    variable and then pass it.

Author: ```suggestion
   consequently, avoid directly passing a function call or some other
   operation with side-effects to this macro. Instead, store the result as a
   variable and then pass it.
```

   .. versionadded:: 3.3

.. c:macro:: Py_ALIGNED(num)

   Specify alignment to *num* bytes on compilers that support it.

   This does not use the C11 standard ``_Alignas`` specifier; prefer use of

>    that over this macro.

Reviewer: ```suggestion
   Consider using the C11 standard ``_Alignas`` specifier over this macro.
```

.. c:macro:: Py_ARITHMETIC_RIGHT_SHIFT(type, integer, positions)

   Similar to :c:expr:`integer >> positions`, but forces sign extension, as the C
   standard does not define whether a right-shift of a signed integer will
   perform sign extension or a zero-fill.

   *integer* should be any signed integer type.
   *positions* is the number of positions to shift to the right.

   Both *integer* and *positions* can be evaluated more than once;
   consequently, avoid directly passing a function call or some other
   expensive operation to this macro. Instead, store the result as a
   variable and then pass it.

   *type* is the type of the expression. In modern versions of Python, this is

>    unused. Historically, *type* was used to cast *integer*.

Reviewer: Can we reformulate this more precisely through a versionchanged? And indicate that the parameter is unused and kept for backward compatibility?

Author: There's already a `versionchanged` that specifies this below.

Reviewer: Oh sorry. In this case, I think we can just say that it's not used anymore and only kept for BC. The change has been done in 3.1 so whatever is after 3.1 is already "modern" Python.

Author: Done.

   .. versionadded:: 3.11

.. c:macro:: Py_SAFE_DOWNCAST(value, larger, smaller)

   Cast *value* to type *smaller* from type *larger*, validating that no
   information was lost.

   On release builds of Python, this is roughly equivalent to

>    :c:expr:`(smaller) value` (in C++, ``static_cast<smaller>(value)`` will be

Reviewer 2: It emits a warning:

```
c:identifier reference target not found: smaller [ref.identifier]
```

I suggest to avoid c:expr:

```suggestion
   ``(smaller) value`` (in C++, ``static_cast<smaller>(value)`` will be
```

   used instead).

   On debug builds (implying that :c:macro:`Py_DEBUG` is defined), this asserts
   that no information was lost with the cast from *larger* to *smaller*.

   *value*, *larger*, and *smaller* may all be evaluated more than once in the
   expression; consequently, do not pass an expensive operation directly to

>    this macro.

Author: ```suggestion
   expression; consequently, do not pass an expression with side-effects directly to
   this macro.
```

>    this macro.

Reviewer 2: Expressions with side effects (ex: ``value++``) should also be avoided.

   .. versionadded:: 3.11

.. c:macro:: Py_SAFE_DOWNCAST(value, larger, smaller)

   Cast *value* to type *smaller* from type *larger*, validating that no
   information was lost.

   On release builds of Python, this is roughly equivalent to
   :c:expr:`(smaller) value` (in C++, ``static_cast<smaller>(value)`` will be
   used instead).

   On debug builds (implying that :c:macro:`Py_DEBUG` is defined), this asserts
   that no information was lost with the cast from *larger* to *smaller*.

   *value*, *larger*, and *smaller* may all be evaluated more than once in the

>    expression; as a consequence, do not pass an expensive operation directly to

Reviewer: ```suggestion
   expression; consequently, do not pass an expensive operation directly to
```

   this macro.

.. c:macro:: Py_STRINGIFY(x)

   Convert ``x`` to a C string.  E.g. ``Py_STRINGIFY(123)`` returns
   ``"123"``.

   .. versionadded:: 3.4

.. c:macro:: Py_ULL(number)

   Similar to :c:macro:`Py_LL`, but *number* will be an ``unsigned long long``
   literal instead. This is done by appending ``U`` to the result of ``Py_LL``.

   In modern versions of Python, this macro is not very useful, as C99 and

>    later require the ``ULL``/``LLU`` suffix to be valid for an integer.

Reviewer: ```suggestion
   later require the ``ULL``/``LLU`` suffixes to be valid for an integer.
```
